File: SAAF/java_template/src/main/java/lambda/Sale.java
package lambda;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.temporal.ChronoUnit;
import java.util.Objects;

public class Sale {
    private static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("M/d/yyyy");

    String region;
    String country;
    String itemType;
    String salesChannel;
    String orderPriority;
    LocalDate orderDate;
    int orderId;
    LocalDate shipDate;
    int unitsSold;
    BigDecimal unitPrice;
    BigDecimal unitCost;
    BigDecimal totalRevenue;
    BigDecimal totalCost;
    BigDecimal totalProfit;

    // New Variables
    int orderProcessingTime;
    BigDecimal grossMargin;

    public Sale() {
    }

    public Sale(String region, String country, String itemType, String salesChannel, String orderPriority, LocalDate orderDate, int orderId, LocalDate shipDate, int unitsSold, BigDecimal unitPrice, BigDecimal unitCost, BigDecimal totalRevenue, BigDecimal totalCost, BigDecimal totalProfit) {
        this.region = region;
        this.country = country;
        this.itemType = itemType;
        this.salesChannel = salesChannel;

        this.orderPriority = formatPriority(orderPriority);

        this.orderDate = orderDate;
        this.orderId = orderId;
        this.shipDate = shipDate;
        this.unitsSold = unitsSold;
        this.unitPrice = unitPrice;
        this.unitCost = unitCost;
        this.totalRevenue = totalRevenue;
        this.totalCost = totalCost;
        this.totalProfit = totalProfit;

        if (orderDate != null && shipDate != null) {
            this.orderProcessingTime = Math.toIntExact(ChronoUnit.DAYS.between(orderDate, shipDate));
        }

        this.grossMargin = totalProfit.divide(totalRevenue, 2, RoundingMode.HALF_UP);
    }

    public String getRegion() {
        return region;
    }

    public void setRegion(String region) {
        this.region = region;
    }

    public String getCountry() {
        return country;
    }

    public void setCountry(String country) {
        this.country = country;
    }

    public String getItemType() {
        return itemType;
    }

    public void setItemType(String itemType) {
        this.itemType = itemType;
    }

    public String getSalesChannel() {
        return salesChannel;
    }

    public void setSalesChannel(String salesChannel) {
        this.salesChannel = salesChannel;
    }

    public String getOrderPriority() {
        return orderPriority;
    }

    public void setOrderPriority(String orderPriority) {
        this.orderPriority = orderPriority;
    }

    public LocalDate getOrderDate() {
        return orderDate;
    }

    public void setOrderDate(LocalDate orderDate) {
        this.orderDate = orderDate;
    }

    public int getOrderId() {
        return orderId;
    }

    public void setOrderId(int orderId) {
        this.orderId = orderId;
    }

    public LocalDate getShipDate() {
        return shipDate;
    }

    public void setShipDate(LocalDate shipDate) {
        this.shipDate = shipDate;
    }

    public int getUnitsSold() {
        return unitsSold;
    }

    public void setUnitsSold(int unitsSold) {
        this.unitsSold = unitsSold;
    }

    public BigDecimal getUnitPrice() {
        return unitPrice;
    }

    public void setUnitPrice(BigDecimal unitPrice) {
        this.unitPrice = unitPrice;
    }

    public BigDecimal getUnitCost() {
        return unitCost;
    }

    public void setUnitCost(BigDecimal unitCost) {
        this.unitCost = unitCost;
    }

    public BigDecimal getTotalRevenue() {
        return totalRevenue;
    }

    public void setTotalRevenue(BigDecimal totalRevenue) {
        this.totalRevenue = totalRevenue;
    }

    public BigDecimal getTotalCost() {
        return totalCost;
    }

    public void setTotalCost(BigDecimal totalCost) {
        this.totalCost = totalCost;
    }

    public BigDecimal getTotalProfit() {
        return totalProfit;
    }

    public void setTotalProfit(BigDecimal totalProfit) {
        this.totalProfit = totalProfit;
    }

    @Override
    public boolean equals(Object o) {
        if (o == null || getClass() != o.getClass()) return false;
        Sale sale = (Sale) o;
        return orderId == sale.orderId;
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(orderId);
    }

    @Override
    public String toString() {
        return region + "," +
                country + "," +
                itemType + "," +
                salesChannel + "," +
                orderPriority + "," +
                orderDate.format(FORMATTER) + "," +
                orderId + "," +
                shipDate.format(FORMATTER) + "," +
                unitsSold + "," +
                unitPrice + "," +
                unitCost + "," +
                totalRevenue + "," +
                totalCost + "," +
                totalProfit + "," +
                orderProcessingTime + "," +
                grossMargin;
    }

    private String formatPriority(String priority) {
        if (priority != null) {
            switch (priority) {
                case "L":
                    return "Low";
                case "M":
                    return "Medium";
                case "H":
                    return "High";
                case "C":
                    return "Critical";
            }
        }

        return "";
    }
}

File: SAAF/python_template/src/sale.py
from datetime import datetime
from decimal import Decimal, ROUND_HALF_UP

DATE_FORMAT = "%m/%d/%Y"


class Sale:
    def __init__(
        self,
        region,
        country,
        item_type,
        sales_channel,
        order_priority,
        order_date,
        order_id,
        ship_date,
        units_sold,
        unit_price,
        unit_cost,
        total_revenue,
        total_cost,
        total_profit
    ):
        self.region = region
        self.country = country
        self.item_type = item_type
        self.sales_channel = sales_channel
        self.order_priority = self.format_priority(order_priority)

        self.order_date = order_date
        self.order_id = order_id
        self.ship_date = ship_date
        self.units_sold = units_sold

        self.unit_price = unit_price
        self.unit_cost = unit_cost
        self.total_revenue = total_revenue
        self.total_cost = total_cost
        self.total_profit = total_profit

        if self.order_date and self.ship_date:
            self.order_processing_time = (self.ship_date - self.order_date).days
        else:
            self.order_processing_time = 0

        self.gross_margin = (
            self.total_profit / self.total_revenue
        ).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)

    def format_priority(self, priority):
        return {
            "L": "Low",
            "M": "Medium",
            "H": "High",
            "C": "Critical"
        }.get(priority, "")

    def __eq__(self, other):
        return isinstance(other, Sale) and self.order_id == other.order_id

    def __hash__(self):
        return hash(self.order_id)

    def __str__(self):
        return ",".join([
            self.region,
            self.country,
            self.item_type,
            self.sales_channel,
            self.order_priority,
            self.order_date.strftime(DATE_FORMAT),
            str(self.order_id),
            self.ship_date.strftime(DATE_FORMAT),
            str(self.units_sold),
            str(self.unit_price),
            str(self.unit_cost),
            str(self.total_revenue),
            str(self.total_cost),
            str(self.total_profit),
            str(self.order_processing_time),
            str(self.gross_margin)
        ])